src/main/java/com/arpnetworking/commons/hostresolver/BackgroundCachingHostResolver.java

```diff
@@ -0,0 +1,106 @@
+
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+
+import java.net.UnknownHostException;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Caches results of a <code>HostResolver</code> and refresh it in the
+ * background to avoid any blocking.
+ *
+ * Dependencies:
+ * <ul>
+ *     <li><i>None</i></li>
+ * </ul>
+ *
+ * @author Ville Koskela (ville dot koskela at inscopemetrics dot com)
+ */
+public class BackgroundCachingHostResolver implements HostResolver {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getLocalHostName() throws UnknownHostException {
+        if (!_cachedLocalHostName.isPresent()) {
+            throw new UnknownHostException();
+        }
+        return _cachedLocalHostName.get();
+    }
+
+    /**
+     * Constructor for a <code>BackgroundCachingHostResolver</code> wrapping the <code>DefaultHostResolver</code>.
+     *
+     * @param rate The time period in which to execute one background refresh.
+     */
+    public BackgroundCachingHostResolver(final Duration rate) {
+        this(DEFAULT_WRAPPED_HOST_RESOLVER, rate);
+    }
+
+    /**
+     * Constructor for a <code>BackgroundCachingHostResolver</code> wrapping the specified <code>HostResolver</code>.
+     *
+     * @param wrappedHostResolver The <code>HostResolver</code> to delegate to periodically.
+     * @param rate The time period in which to execute one background refresh.
+     */
+    public BackgroundCachingHostResolver(final HostResolver wrappedHostResolver, final Duration rate) {
+        _wrappedHostResolver = wrappedHostResolver;
+        _cachedLocalHostName = Optional.empty();
+
+        updateHost();
+
+        _scheduledExecutorService = new ScheduledThreadPoolExecutor(
+                1,
+                runnable -> {
+                    final Thread thread = new Thread(runnable, "BackgroundCachingHostResolver");
+                    thread.setDaemon(true);
+                    return thread;
+                });
+
+        _scheduledExecutorService.scheduleAtFixedRate(
+                this::updateHost,
+                rate.toMillis(),
+                rate.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    /* package private */ void updateHost() {
+        try {
+            _cachedLocalHostName = Optional.ofNullable(_wrappedHostResolver.getLocalHostName());
+        } catch (final UnknownHostException e) {
+            LOGGER.warn()
+                    .setMessage("Unable to resolve host")
+                    .setThrowable(e)
+                    .log();
+        }
+    }
+
+    private final ScheduledExecutorService _scheduledExecutorService;
+    private final HostResolver _wrappedHostResolver;
+
+    private Optional<String> _cachedLocalHostName;
```

Comment:
I think this needs to be thread safe.  Ya know, since you're actually the one using it in a multi-threaded way. ( ;